hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/Planner.java

```diff
@@ -101,13 +109,109 @@ DAG createDag() {
+                originalParents.put(downstream, new ArrayList<>(downstream.upstream()));
+                downstream.upstream().replaceAll(p -> p == lastInChain ? fused : p);
+            }
+        }
+
+        for (Transform transform : transforms) {
             transform.addToDag(this);
         }
+
+        // restore original parents
+        for (Entry<Transform, List<Transform>> en : originalParents.entrySet()) {
+            List<Transform> upstream = en.getKey().upstream();
+            for (int i = 0; i < upstream.size(); i++) {
+                en.getKey().upstream().set(i, en.getValue().get(i));
+            }
+        }
+
         return dag;
     }
 
+    private static List<Transform> findFusableChain(Transform transform, Map<Transform, List<Transform>> adjacencyMap) {
+        ArrayList<Transform> chain = new ArrayList<>();
+        for (;;) {
+            if (transform instanceof MapTransform || transform instanceof FlatMapTransform) {
+                chain.add(transform);
+                List<Transform> downstreams = adjacencyMap.get(transform);
+                if (downstreams.size() == 1 && downstreams.get(0).localParallelism() == transform.localParallelism()) {
+                    transform = downstreams.get(0);
+                    continue;
+                }
+            }
+            break;
+        }
+        return chain.size() > 1 ? chain : null;
+    }
+
+    private static Transform fuseFlatMapTransforms(List<Transform> chain) {
+        assert chain.size() > 1 : "chain.size()=" + chain.size();
+        assert chain.get(0).upstream().size() == 1;
+
+        int lastFlatMap = 0;
+        FunctionEx<Object, Traverser> flatMapFn = null;
+        for (int i = 0; i < chain.size(); i++) {
+            if (chain.get(i) instanceof FlatMapTransform) {
+                FunctionEx<Object, Traverser> function = ((FlatMapTransform) chain.get(i)).flatMapFn();
+                FunctionEx<Object, Object> inputMapFn = mergeMapFunctions(chain.subList(lastFlatMap, i));
+                if (inputMapFn != null) {
+                    flatMapFn = flatMapFn == null
+                            ? (Object t) -> {
+                                Object mappedValue = inputMapFn.apply(t);
+                                return mappedValue != null ? function.apply(mappedValue) : Traversers.empty();
+                            }
+                            : flatMapFn.andThen(r -> r.map(inputMapFn).flatMap(function));
+                } else {
+                    flatMapFn = flatMapFn == null
+                            ? function::apply
+                            : flatMapFn.andThen(r -> r.flatMap(function));
+                }
+                lastFlatMap = i + 1;
+            }
+        }
+
+        FunctionEx trailingMapFn = mergeMapFunctions(chain.subList(lastFlatMap, chain.size()));
+        String name = chain.stream().map(Transform::name).collect(Collectors.joining(", ", "fused(", ")"));
```

Comment:
this could create quite long names, how will look in the Management Center?

Reply:
It's an issue of MC imo, it should trim it somehow maybe using `text-overflow:ellipsis` :)
Or the user, seeing it, can rename the the stages. The names can be user-assigned, i didn't want to just replace them or not display all of them...